nexus/api.go

```diff
@@ -176,3 +203,82 @@ func addContextTimeoutToHTTPHeader(ctx context.Context, httpHeader http.Header)
 	httpHeader.Set(HeaderRequestTimeout, time.Until(deadline).String())
 	return httpHeader
 }
+
+// Link contains an URL and a Type that can be used to decode the URL.
+// Links can contain any arbitrary information as a percent-encoded URL.
+// It can be used to pass information about the caller to the handler, or vice-versa.
+type Link struct {
+	// URL information about the link.
+	// It must be URL percent-encoded.
+	URL string
+	// Type can describe an actual data type for decoding the URL.
+	// Valid chars: alphanumeric, '_', '.', '/'
+	Type string
+}
+
+const linkTypeKey = "type"
+
+func encodeLink(link Link) (string, error) {
+	u, err := url.Parse(link.URL)
+	if err != nil {
+		return "", fmt.Errorf("failed to encode link: url not valid: %s", link.URL)
+	}
+	for _, p := range strings.Split(u.EscapedPath(), "/") {
+		unescaped, err := url.PathUnescape(p)
+		if err != nil || p != url.PathEscape(unescaped) {
+			return "", fmt.Errorf("failed to encode link: path not percent-encoded: %s", link.URL)
+		}
+	}
+	_, err = url.ParseQuery(u.RawQuery)
+	if err != nil {
+		return "", fmt.Errorf("failed to encode link: query not percent-encoded: %s", link.URL)
+	}
+
+	validSymbols := []rune("_./")
+	for _, c := range link.Type {
+		if !unicode.IsLetter(c) && !unicode.IsDigit(c) && !slices.Contains(validSymbols, c) {
+			return "", fmt.Errorf("failed to encode link: type contains invalid chars: %s", link.Type)
+		}
+	}
+
+	return fmt.Sprintf(`<%s>; %s="%s"`, link.URL, linkTypeKey, link.Type), nil
+}
+
+func decodeLink(encodedLink string) (Link, error) {
+	var link Link
+	parts := strings.Split(encodedLink, ";")
+
+	u := strings.TrimSpace(parts[0])
+	if !strings.HasPrefix(u, "<") || !strings.HasSuffix(u, ">") {
+		return link, fmt.Errorf("failed to parse link header value: %s", encodedLink)
+	}
+	u = u[1 : len(u)-1]
+	if _, err := url.Parse(u); err != nil {
+		return link, fmt.Errorf("failed to parse link header value: %s", encodedLink)
+	}
+	link.URL = u
+
+	for _, part := range parts[1:] {
+		part = strings.TrimSpace(part)
+		if len(part) == 0 {
+			continue
+		}
+		kv := strings.SplitN(part, "=", 2)
+		if len(kv) != 2 {
+			return link, fmt.Errorf("failed to parse link header value: %s", encodedLink)
+		}
+		key := strings.TrimSpace(kv[0])
+		val := strings.TrimSpace(kv[1])
+		if strings.HasPrefix(val, `"`) != strings.HasSuffix(val, `"`) {
+			return link, fmt.Errorf("failed to parse link header value: %s", encodedLink)
+		}
+		if strings.HasPrefix(val, `"`) {
+			val = val[1 : len(val)-1]
+		}
+		if strings.ToLower(key) == linkTypeKey {
```

Comment:
```suggestion
		if key == linkTypeKey {
```

I don't think this needs to be case insensitive. Also, arguably you should validate the value for type too before returning a potentially bad link that will encoding. Consider making the URL and type validators utilities you call on encode and decode.